golem-registry-service/src/repo/environment_share.rs

```diff
@@ -0,0 +1,417 @@
+            .fetch_one_as(
+                sqlx::query_as(indoc! { r#"
+                    INSERT INTO environment_share_revisions
+                    (environment_share_id, revision_id, created_at, created_by, deleted)
+                    VALUES ($1, $2, $3, $4, $5)
+                    RETURNING environment_share_id, revision_id, created_at, created_by, deleted
+                "# })
+                .bind(revision.environment_share_id)
+                .bind(revision.revision_id)
+                .bind_deletable_revision_audit(revision.audit),
+            )
+            .await
+            .to_error_on_unique_violation(EnvironmentShareRepoError::RevisionAlreadyExists {
+                revision_id: revision.revision_id,
+            })?;
+
+        revision.roles = {
+            let mut inserted_roles = Vec::with_capacity(revision.roles.len());
+
+            for role in original_roles {
+                let role = role
+                    .ensure_environment_share(revision.environment_share_id, revision.revision_id);
+                inserted_roles.push(Self::insert_one_role(tx, role).await?);
+            }
+
+            // sort roles so they are consistent with get_roles ordering
+            inserted_roles.sort_by_key(|r| r.role);
+
+            inserted_roles
+        };
+
+        Ok(revision)
+    }
+}
+
+#[trait_gen(PostgresPool -> PostgresPool, SqlitePool)]
+#[async_trait]
+impl EnvironmentShareRepo for DbEnvironmentShareRepo<PostgresPool> {
+    async fn create(
+        &self,
+        environment_id: Uuid,
+        revision: EnvironmentShareRevisionRecord,
+        grantee_account_id: Uuid,
+    ) -> Result<EnvironmentShareExtRevisionRecord, EnvironmentShareRepoError> {
+        let revision = revision.ensure_first();
+
+        self.db_pool.with_tx_err(METRICS_SVC_NAME, "create", |tx| {
+            async move {
+                let environment_share_record: EnvironmentShareRecord = tx
+                    .fetch_one_as(
+                        sqlx::query_as(indoc! {r#"
+                            INSERT INTO environment_shares (environment_id, environment_share_id, grantee_account_id, created_at, updated_at, deleted_at, modified_by, current_revision_id)
+                            VALUES ($1, $2, $3, $4, $4, NULL, $5, $6)
+                            RETURNING environment_id, environment_share_id, grantee_account_id, created_at, updated_at, deleted_at, modified_by, current_revision_id
+                        "#})
+                            .bind(environment_id)
+                            .bind(revision.environment_share_id)
+                            .bind(grantee_account_id)
+                            .bind(&revision.audit.created_at)
+                            .bind(revision.audit.created_by)
+                            .bind(revision.revision_id)
+                    )
+                    .await
+                    .to_error_on_unique_violation(EnvironmentShareRepoError::ShareViolatesUniqueness)?;
```

Comment:
i would simply consider this concurrent modification error, and do not make more specific ones, like RevisionForUpdateNotFound in update (which is kinda also misleading, it is not "not found", rather, not the current one). just like with the prev repos i would like to keep them ConcurrentModification errors, as the clients should behave the same: refetch their data, and then decide what to do.

it is good to have more specific errors for user provided identitiy uniqueness errors (version, name, email), but for optimistic lock errors i do not think that client can do anything meaningfully different, and depending on interleave patterns they might get one or the other.